sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/StructuredConcurrencyPrinciplesCheck.kt

```diff
@@ -0,0 +1,73 @@
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.ASYNC
+import org.sonarsource.kotlin.api.CallAbstractCheck
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.KOTLINX_COROUTINES_PACKAGE
+import org.sonarsource.kotlin.api.LAUNCH
+import org.sonarsource.kotlin.api.WITH_CONTEXT
+import org.sonarsource.kotlin.api.predictReceiverExpression
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val JOB_CONSTRUCTOR = FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE, name = "Job")
+private val SUPERVISOR_JOB_CONSTRUCTOR = FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE, name = "SupervisorJob")
+private const val MESSAGE_ENDING = " here leads to breaking of structured concurrency principles."
+private const val DELICATE_API_CLASS_TYPE = "kotlin.reflect.KClass<kotlinx.coroutines.DelicateCoroutinesApi>"
+
+@Rule(key = "S6306")
+class StructuredConcurrencyPrinciplesCheck : CallAbstractCheck() {
+
+    override fun functionsToVisit() = listOf(
+        FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE, name = ASYNC),
+        FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE, name = LAUNCH),
+        FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE, name = WITH_CONTEXT),
+    )
+
+    override fun visitFunctionCall(callExpression: KtCallExpression, resolvedCall: ResolvedCall<*>, kotlinFileContext: KotlinFileContext) {
+        val bindingContext = kotlinFileContext.bindingContext
+        val receiver = callExpression.predictReceiverExpression(bindingContext) as? KtNameReferenceExpression
+        if (receiver?.getReferencedName() == "GlobalScope" && !callExpression.checkOptInDelicateApi(bindingContext)) {
+            kotlinFileContext.reportIssue(receiver, """Using "GlobalScope"$MESSAGE_ENDING""")
+        } else {
+            resolvedCall.valueArgumentsByIndex?.let { args ->
+                if (args.isEmpty()) return
+                val argExprCall = (args[0] as? ExpressionValueArgument)?.valueArgument?.getArgumentExpression() as? KtCallExpression ?: return
+                if (JOB_CONSTRUCTOR.matches(argExprCall, bindingContext) || SUPERVISOR_JOB_CONSTRUCTOR.matches(argExprCall, bindingContext)
+                ) {
+                    kotlinFileContext.reportIssue(argExprCall, """Using "${argExprCall.text}"$MESSAGE_ENDING""")
+                }
+            }
+        }
+    }
+
+    private fun KtExpression.checkOptInDelicateApi(bindingContext: BindingContext): Boolean {
+        var parent: PsiElement? = this
+        while (parent != null) {
+            val annotations = (parent as? KtAnnotated)?.annotationEntries
+            if (annotations.isAnnotatedWithOptInDelicateApi(bindingContext)) return true
+            parent = parent.parent
+        }
+        return false
+    }
+}
+
+private fun MutableList<KtAnnotationEntry>?.isAnnotatedWithOptInDelicateApi(bindingContext: BindingContext) = this?.let {
+    it.any { annotation ->
+        annotation.shortName?.asString() == "OptIn"
```

Comment:
Another way to opt into the delicate coroutines API is using the `@DelicateCoroutinesApi` annotation, currently we don't check this here.